sdk/ledger-service/http-json/src/test/scala/com/digitalasset/http/util/LockSetTest.scala

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.http.util
+
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+import java.util.concurrent.ConcurrentHashMap
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+
+import scala.concurrent.{ExecutionContext, Future, Await}
+import scala.concurrent.duration.DurationInt
+import scala.util.Random
+import scala.jdk.CollectionConverters._
+
+class LockSetTest extends AnyFlatSpec with Matchers {
+
+  val logger = ContextualizedLogger.get(getClass)
+  implicit val ec: ExecutionContext = ExecutionContext.global
+  implicit val lc: LoggingContext = LoggingContext.empty
+
+  "withLocksOn" should "synchronize without deadlock" in {
+    val lockSet = new LockSet[Char](logger)
+    val keys = ('a' to 'z').toSet
+    val counters = new ConcurrentHashMap[Char, Long]()
+
+    val numThreads = 100 // Concurrently executing threads updating counters
+    val repetitions = 20 // How many times the step is repeated
+    val numKeysToIncr = 5 // How many counters we attempt to increment each step
+
+    val futures = (1 to numThreads).map { _ =>
+      Future {
+        (1 to repetitions).foreach { i =>
+          val keysToIncr = Random.shuffle(keys.toVector).take(numKeysToIncr)
+          val task = lockSet
+            .withLocksOn(keysToIncr) {
+              Future {
+                keysToIncr.foreach { k =>
+                  counters.put(k, counters.getOrDefault(k, 0L) + 1) // Intentially racy.
+                }
+                if (i % 2 == 0) throw new RuntimeException("ensure we can handle exceptions")
```

Comment:
Can you use a custom exception and catch only that one. There could be other runtime exceptions thrown that you don't want to inadvertently catch.